substrate/frame/election-provider-support/src/lib.rs

```diff
@@ -674,10 +896,303 @@ pub type Voter<AccountId, Bound> = (AccountId, VoteWeight, BoundedVec<AccountId,
+	fn clone(&self) -> Self {
+		Self { voters: self.voters.clone(), total: self.total }
+	}
+}
+
+impl<AccountId, Bound: Get<u32>> TryFrom<sp_npos_elections::Support<AccountId>>
+	for BoundedSupport<AccountId, Bound>
+{
+	type Error = &'static str;
+	fn try_from(s: sp_npos_elections::Support<AccountId>) -> Result<Self, Self::Error> {
+		let voters = s.voters.try_into().map_err(|_| "voters bound not respected")?;
+		Ok(Self { voters, total: s.total })
+	}
+}
+
+impl<AccountId: Clone, Bound: Get<u32>> BoundedSupport<AccountId, Bound> {
+	pub fn sorted_truncate_from(mut support: sp_npos_elections::Support<AccountId>) -> (Self, u32) {
+		// If bounds meet, then short circuit.
+		if let Ok(bounded) = support.clone().try_into() {
+			return (bounded, 0)
+		}
+
+		let pre_len = support.voters.len();
+		// sort support based on stake of each backer, low to high.
+		// Note: we don't sort high to low and truncate because we would have to track `total`
+		// updates, so we need one iteration anyhow.
+		support.voters.sort_by(|a, b| a.1.cmp(&b.1));
+		// then do the truncation.
+		let mut bounded = Self { voters: Default::default(), total: 0 };
+		while let Some((voter, weight)) = support.voters.pop() {
+			if let Err(_) = bounded.voters.try_push((voter, weight)) {
+				break
+			}
+			bounded.total += weight;
+		}
+		let post_len = bounded.voters.len();
+		(bounded, (pre_len - post_len) as u32)
+	}
+}
+
+/// A bounded vector of [`BoundedSupport`].
+///
+/// A [`BoundedSupports`] is a set of [`sp_npos_elections::Supports`] which are bounded in two
+/// dimensions. `BInner` corresponds to the bound of the maximum backers per voter and `BOuter`
+/// corresponds to the bound of the maximum winners that the bounded supports may contain.
+///
+/// With the bounds, we control the maximum size of a bounded supports instance.
+#[derive(Encode, Decode, DecodeWithMemTracking, TypeInfo, DefaultNoBound, MaxEncodedLen)]
+#[codec(mel_bound(AccountId: MaxEncodedLen, BOuter: Get<u32>, BInner: Get<u32>))]
+#[scale_info(skip_type_params(BOuter, BInner))]
+pub struct BoundedSupports<AccountId, BOuter: Get<u32>, BInner: Get<u32>>(
+	pub BoundedVec<(AccountId, BoundedSupport<AccountId, BInner>), BOuter>,
+);
+
+/// Try and build yourself from another `BoundedSupports` with a different set of types.
+pub trait TryFromOtherBounds<AccountId, BOtherOuter: Get<u32>, BOtherInner: Get<u32>> {
+	fn try_from_other_bounds(
+		other: BoundedSupports<AccountId, BOtherOuter, BOtherInner>,
+	) -> Result<Self, crate::Error>
+	where
+		Self: Sized;
+}
+
+impl<
+		AccountId,
+		BOuter: Get<u32>,
+		BInner: Get<u32>,
+		BOtherOuter: Get<u32>,
+		BOuterInner: Get<u32>,
+	> TryFromOtherBounds<AccountId, BOtherOuter, BOuterInner>
+	for BoundedSupports<AccountId, BOuter, BInner>
+{
+	fn try_from_other_bounds(
+		other: BoundedSupports<AccountId, BOtherOuter, BOuterInner>,
+	) -> Result<Self, crate::Error> {
+		// NOTE: we might as well do this with unsafe rust and do it faster.
```

Comment:
I spotted an issue in this in a pretty late review. TODO for me: add a couple of tests